extensions/README.md

```diff
@@ -0,0 +1,152 @@
+# Blade Extensions
+
+There are several ways that you can extend `blade` CLI.  Most simply you can add custom `commands` and `templates`.
+But likely if you want to customize a developer's experience with `blade` to do enable some new developer workflow, you
+want to create custom commands and templates that all are associated with a `blade` _profile_.
+
+When blade cli starts it will look in the user's home directory ${user.home}/.blade/extensions folder for any jar files.
+All jar files will be searched to see if they contain valid blade extensions detailed below.
+
+## Installing new extensions
+
+If you would like to install an existing blade extension you can use a new command. `blade extensions install` to install from one of the following:
+
+1. Local jar file
+2. Remote jar file indicated by a URL
+3. Github repository that contains an extension.
+
+**Examples of how to install extensions**
+
+```
+$ blade extension install /path/to/my_extension.jar
+```
+
+```
+$ blade extension install https://repository.lifera.com/public/group/com/liferay/blade/com.liferay.blade.extension/1.0.0/com.liferay.blade.extension-1.0.0.jar
+```
+
+```
+$ blade extension install https://github.com/gamerson/blade-sample-command
+```
+
+## Developing new extensions
+
+There are a few use-cases that blade extenders probably want to consider.  If they simply want a new `command` that they
+can add to blade's standard set of commands that would apply globally to all types of workspaces, they can simply
+add a new custom command.  See Custom Commands section for details.
+
+Most likely you will have a set of custom commands that only apply to a specific workspace environment.  So lets say
+that you wanted to build some custom commands that only work when the Liferay workspace is configured to use `docker`
+, `wedeploy`, or `openshift`  Then likely what you want is a `profile` that supports the development workflow in that
+environment.  See the section on Profiles down below for more details.
+
+### Custom commands
+
+To create a custom command you need to build a Command class using the JCommander framework.  Then you register your
+new command with blade by using standard JRE service loader mechanisms.  Blade looks for custom commands using
+the this service interfaces: `com.liferay.blade.cli.command.BaseCommand`
+
+Therefore in your extension jar you should have a file like this: `META-INF/services/com.liferay.blade.cli.command.BaseCommand1` and inside would be a list of all of your FQDN classes of your custom commands.
+
+An example of how this should be laid out can be found in the blade repository.
+
+https://github.com/liferay/liferay-blade-cli/tree/master/extensions/sample-command
```

Comment:
Should we update the dependencies section of this project's `build.gradle` to not use project-level dependencies (`compileOnly project(":cli")`, etc), but rather `compileOnly group: "com.liferay.blade", name: "com.liferay.blade.cli", version: "latest.release"`
Or perhaps we should point the user to https://github.com/gamerson/blade-sample-command/ instead, which is more of a standalone project?